File: Data Structures & Algorithms/search-2d-matrix/submission-2.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        int rowst=0,rowend=rows-1;
        int colst=0,colsend=cols-1;
        while(rowst<=rowend && colst<=colsend){
            int rowmid = (rowst+rowend)/2;
            int colmid = (colst+colsend)/2;

            if(matrix[rowmid][colmid]==target){
                return true;
            }else if(matrix[rowmid][colmid]<target){
                if(matrix[rowmid][cols-1]<target){
                    rowst++;
                }else{
                    colst++;
                }
            }else{
                if(matrix[rowmid][0]>target){
                    rowend--;
                }else{
                    colsend--;
                }
            }
        }
        return false;
    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-2.java
class Solution {
    public void bfsHelper(char[][] grid,int i,int j,int rows,int cols){
        if(i<0 || j<0 || i>=rows || j>=cols || grid[i][j]=='0'){
            return;
        }
        grid[i][j]='0';
        bfsHelper(grid,i+1,j,rows,cols);
        bfsHelper(grid,i-1,j,rows,cols);
        bfsHelper(grid,i,j+1,rows,cols);
        bfsHelper(grid,i,j-1,rows,cols);
    }
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int ni = 0;
        for(int i=0;i<rows;i++){
            for(int j=0;j<cols;j++){
                if(grid[i][j]=='1'){
                    ni++;
                    bfsHelper(grid,i,j,rows,cols);
                }
            }
        }
        return ni;
    }
}
